src/Data/List/Sufficient.agda

```diff
@@ -0,0 +1,87 @@
+------------------------------------------------------------------------
+-- The Agda standard library
+--
+-- 'Sufficient' lists: a structurally inductive view of lists xs
+-- as given by xs ≡ non-empty prefix + sufficient suffix
+--
+-- Useful for termination arguments for function definitions
+-- which provably consume a non-empty (but otherwise arbitrary) prefix
+-- *without* having to resort to ancillary WF induction on length etc.
+-- e.g. lexers, parsers etc.
+--
+-- Credited by Conor McBride as originally due to James McKinna
+------------------------------------------------------------------------
+
+{-# OPTIONS --without-K --safe #-}
+
+module Data.List.Sufficient {a} {A : Set a} where
```

Comment:
So I tried this... and the very first definition of `suffAcc` breaks:
```agda
private
  variable

    a : Level
    A : Set a
    x : A
    xs : List A

------------------------------------------------------------------------
-- Sufficient builder

suffAcc : ∀ {b} (B :  List A → Set b) (xs : List A) → Set (a ⊔ b)
suffAcc B xs = ∀ {x} {prefix} suffix → xs ≡ x ∷ prefix ++ suffix → B suffix
```
yielding
```agda
Set (A.a ⊔ b) != Set (a ⊔ b)
when checking that the expression
∀ {x} {prefix} suffix → xs ≡ x ∷ prefix ++ suffix → B suffix has
type Set (a ⊔ b)
```
and I realise that I *hate* these kinds of errors (which arise in my experience  precisely, and only, from using `variable` declarations... :-() because I just don't understand how I am supposed to fix them (because they make no sense: surely `suffAcc` is defined within their scope, so that `a`, `A`etc. are also in scope ...

... so some Agda hand-holding would be welcome here!

Reply:
In that particular case you'd need to explicitly have `{A : Set a}`. The hidden `a` generated by `A` is distinct from the `a` you explicitly wrote, I guess so that you can alpha-rename things without them breaking mysteriously. https://agda.readthedocs.io/en/v2.6.2.2/language/generalization-of-declared-variables.html explains them in detail.

Reply:
Thanks @Taneb I had just re-read the explanation in detail, and I still don't have an adequate internal mental model of what it's telling me... and in particular *why* I need the additional outermost parametric declaration... but commit `a53300d` seems to do the trick (but with some additional knock-on effects, sigh)